MatrixKit/Models/Contact/MXKContactManager.m

```diff
@@ -1230,74 +1229,125 @@ + (void)requestUserConfirmationForLocalContactsSyncWithTitle:(NSString*)title
+        NSArray *mxUsers = mxSession.users;
+        
+        for (MXUser *user in mxUsers)
+        {
+            // Check whether this user has already been added
+            if (!matrixContactByMatrixID[user.userId])
+            {
+                if ((self.contactManagerMXRoomSource == MXKContactManagerMXRoomSourceAll) || ((self.contactManagerMXRoomSource == MXKContactManagerMXRoomSourceDirectChats) && mxSession.directRooms[user.userId]))
+                {
+                    // Check whether a contact is already defined for this id in previous dictionary
+                    // (avoid delete and create the same ones, it could save thumbnail downloads).
+                    MXKContact* contact = matrixContactByMatrixID[user.userId];
+                    if (contact)
+                    {
+                        contact.displayName = (user.displayname.length > 0) ? user.displayname : user.userId;
+                        
+                        // Check the avatar change
+                        if ((user.avatarUrl || contact.matrixAvatarURL) && ([user.avatarUrl isEqualToString:contact.matrixAvatarURL] == NO))
+                        {
+                            [contact resetMatrixThumbnail];
+                        }
+                    }
+                    else
+                    {
+                        contact = [[MXKContact alloc] initMatrixContactWithDisplayName:((user.displayname.length > 0) ? user.displayname : user.userId) andMatrixID:user.userId];
+                    }
+                    
+                    matrixContactByMatrixID[user.userId] = contact;
+                }
+            }
+        }
+    }
+    
+    return [matrixContactByMatrixID copy];
+}
+
 - (void)refreshMatrixContacts
 {
     NSArray *mxSessions = self.mxSessions;
-    
+
     // Check whether at least one session is available
     if (!mxSessions.count)
     {
         matrixContactByMatrixID = nil;
         matrixContactByContactID = nil;
         [self cacheMatrixContacts];
+
+        [[NSNotificationCenter defaultCenter] postNotificationName:kMXKContactManagerDidUpdateMatrixContactsNotification object:nil userInfo:nil];
     }
     else if (self.contactManagerMXRoomSource != MXKContactManagerMXRoomSourceNone)
     {
-        if (!matrixContactByContactID)
-        {
-            [self loadCachedMatrixContacts];
-        }
-    
-        // The existing dictionary of contacts will be replaced by this one
-        NSMutableDictionary *updatedMatrixContactByMatrixID = [[NSMutableDictionary alloc] initWithCapacity:matrixContactByMatrixID.count];
-        for (MXSession *mxSession in mxSessions)
-        {
-            // Check all existing users
-            NSArray *mxUsers = mxSession.users;
+        MXWeakify(self);
+
+        BOOL shouldFetchLocalContacts = self->matrixContactByContactID == nil;
 
-            for (MXUser *user in mxUsers)
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

Comment:
Why do not use the existing `processingQueue` queue?
I am not sure multiplying queues, threads and thread context switches helps here. If I am wrong, we should apply it everywhere.

Reply:
`dispatch_get_main_queue` gave us a system-defined "background" concurrent queue. Currently we slow down the execution by using the same serial queue for every instructions. But we could use it for the moment to avoid possible race conditions.